cmd/ro/roagent/server.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2012 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package roagent
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+
+	"golang.org/x/crypto/ssh"
+)
+
+const SSHROKey string = "ssh-ro"
+
+var errLocked = errors.New("agent: locked")
+
+// Server wraps an Agent and uses it to implement the agent side of
+// the SSH-agent, wire protocol.
+type server struct {
+	roagent *ROAgent
+}
+
+func parseROSigner(req []byte) (*AddedKey, error) {
+	var k roSignerMsg
+	if err := ssh.Unmarshal(req, &k); err != nil {
+		return nil, err
+	}
+
+	pubKey, err := ssh.ParsePublicKey(k.Pub)
+	if err != nil {
+		return nil, errors.New("received ill-formatted public key")
+	}
+
+	return &AddedKey{
+		PrivateKey: &ROSigner{
+			pub:    pubKey,
+			encKey: k.EncKey,
+		},
+		Comment: k.Comments,
+	}, nil
+}
+
+func (s *server) insertIdentity(req []byte) error {
+	var record struct {
+		Type string `sshtype:"17|25"`
+		Rest []byte `ssh:"rest"`
+	}
+
+	// FIXME what is record?
```

Comment:
I don't understand this comment.

Reply:
It was a note to myself since I didn't understand the purpose of the variable `record` at the time. I'll remove it in the next commit.

Reply:
Is there anything else I should fix? I was going to do this along with other changes, but if everything else looks good I can go ahead and remove this.

Reply:
Everything else seems good, I've just not had the bandwidth to go back through for a final check.